crates/admin/src/cluster_controller/logs_controller.rs

```diff
@@ -945,3 +1120,404 @@ impl LogsController {
         }
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use std::collections::HashSet;
+
+    use enumset::{enum_set, EnumSet};
+    use googletest::prelude::*;
+
+    use restate_types::nodes_config::{LogServerConfig, NodeConfig, Role, StorageState};
+    use restate_types::replicated_loglet::ReplicationProperty;
+    use restate_types::PlainNodeId;
+
+    use super::*;
+
+    #[test]
+    #[should_panic(
+        expected = "not implemented: only node-scoped replication is currently supported"
+    )]
+    fn test_select_log_servers_unsupported_replication_scope() {
+        let nodes: Vec<PlainNodeId> = vec![1.into()];
+        let replication =
+            ReplicationProperty::with_scope(LocationScope::Zone, 1.try_into().unwrap());
+
+        let mut nodes_config = NodesConfiguration::default();
+        nodes_config.upsert_node(node(
+            1,
+            enum_set!(Role::LogServer | Role::Worker),
+            StorageState::Provisioning,
+        ));
+
+        let observed_state = ObservedClusterState {
+            alive_nodes: nodes
+                .iter()
+                .copied()
+                .map(|id| (id, id.with_generation(1)))
+                .collect(),
+            ..Default::default()
+        };
```

Comment:
Introducing some kind of fluent test-only builder for these would greatly reduce the verbosity of the tests. I'll circle back to improve this at some point.